test/common/TestCommonEqEpsilon.hs

```diff
@@ -34,6 +37,134 @@ eqEpsilonRef = unsafePerformIO $ newIORef eqEpsilonDefault
+  assertBool (msg eqEpsilon) (abs (expected-actual) < eqEpsilon)
+  where msg errorMargin = (if null preface then "" else preface ++ "\n") ++
+                           "expected: " ++ show expected ++ "\n but got: " ++ show actual ++
+                           "\n (maximum margin of error: " ++ show errorMargin ++ ")"
+
+----------------------------------------------------------------------------
+-- AssertEqualUpToEpsilon class
+----------------------------------------------------------------------------
+
+class (Fractional z, Ord z, Show z) => AssertEqualUpToEpsilon z a where
+  assertEqualUpToEpsilon :: z -- ^ The error margin (i.e., the epsilon)
+                         -> a -- ^ The expected value
+                         -> a -- ^ The actual value
+                         -> Assertion
+
+instance {-# OVERLAPPABLE #-} (Fractional a, Ord a, Show a) => AssertEqualUpToEpsilon a a where
+  assertEqualUpToEpsilon :: a -> a -> a -> Assertion
+  assertEqualUpToEpsilon = assertEqualUpToEps ""
+
+instance {-# OVERLAPPABLE #-} (AssertEqualUpToEpsilon z a,
+                               AssertEqualUpToEpsilon z b) => AssertEqualUpToEpsilon z (a,b) where
+  assertEqualUpToEpsilon :: z -> (a,b) -> (a,b) -> Assertion
+  assertEqualUpToEpsilon eqEpsilon (e1,e2) (a1,a2) =
+    assertEqualUpToEpsilon eqEpsilon e1 a1 >>
+    assertEqualUpToEpsilon eqEpsilon e2 a2
+
+instance {-# OVERLAPPABLE #-} (AssertEqualUpToEpsilon z a,
+                               AssertEqualUpToEpsilon z b,
+                               AssertEqualUpToEpsilon z c) => AssertEqualUpToEpsilon z (a,b,c) where
+  assertEqualUpToEpsilon :: z -> (a,b,c) -> (a,b,c) -> Assertion
+  assertEqualUpToEpsilon eqEpsilon (e1,e2,e3) (a1,a2,a3) =
+    assertEqualUpToEpsilon eqEpsilon e1 a1 >>
+    assertEqualUpToEpsilon eqEpsilon e2 a2 >>
+    assertEqualUpToEpsilon eqEpsilon e3 a3
+
+instance {-# OVERLAPPABLE #-} (AssertEqualUpToEpsilon z a,
+                               AssertEqualUpToEpsilon z b,
+                               AssertEqualUpToEpsilon z c,
+                               AssertEqualUpToEpsilon z d) => AssertEqualUpToEpsilon z (a,b,c,d) where
+  assertEqualUpToEpsilon :: z -> (a,b,c,d) -> (a,b,c,d) -> Assertion
+  assertEqualUpToEpsilon eqEpsilon (e1,e2,e3,e4) (a1,a2,a3,a4) =
+    assertEqualUpToEpsilon eqEpsilon e1 a1 >>
+    assertEqualUpToEpsilon eqEpsilon e2 a2 >>
+    assertEqualUpToEpsilon eqEpsilon e3 a3 >>
+    assertEqualUpToEpsilon eqEpsilon e4 a4
+
+instance {-# OVERLAPPABLE #-} (Traversable t, AssertEqualUpToEpsilon z a) => AssertEqualUpToEpsilon z (t a) where
+  assertEqualUpToEpsilon :: z -> t a -> t a -> Assertion
+  assertEqualUpToEpsilon eqEpsilon expected actual =
+    assert_list (assertEqualUpToEpsilon eqEpsilon) (asList expected) (asList actual)
+
+instance {-# OVERLAPPABLE #-} (VS.Storable a, AssertEqualUpToEpsilon z a) => AssertEqualUpToEpsilon z (VS.Vector a) where
+  assertEqualUpToEpsilon :: z -> VS.Vector a -> VS.Vector a -> Assertion
```

Comment:
We'd also need `Numeric.LinearAlgebra.Matrix`.

Reply:
Done